### PR TITLE
Correction in the extraction of authors by tag and by json

### PR DESCRIPTION
**Pull request by felipehertzer**
In this correction:

- added 'submitted-by' and 'username' tags to xpath
- the maximum size of the author's name has been increased.
- regex has been added to remove emoji from author names often found on sites like buzzfeed
- added a regex to minify json before running the other regex, was having trouble fetching authors when json formatted.
- added a regex to remove json items like images and organization before searching the author
- reorganized the extract_json function as it was overwriting meta tags with none when no json was found

[qsr](https://bit.ly/3hLqZhQ)
Before this fix: `"author": null`
After this fix: `"author": "Kevin Santos"`

[perthnow](https://bit.ly/2VRZKcZ)
Before this fix: `"author": "NCA NewsWire"`
After this fix: `"author": "Finn McHugh"`

[buzzfeed](https://bzfd.it/3xLkf9k)
Before this fix: `"author": "Hameda Nafiz BuzzFeed Staff"`
After this fix: `"author": "Hameda Nafiz"`

[buzzfeed](https://bzfd.it/36Ib4KP)
Before this fix: `"author": "Olivia ❤️"`
After this fix: `"author": "Olivia Community Contributor"`

[build](https://bit.ly/2UjBiky)
Before this fix: `"author": null`
After this fix: `"author": "Thoams Lane"`

[hunterandbligh](https://bit.ly/3BkG69R)
Before this fix: `none`
After this fix: `"author": "REBECCA MAGRO"`

[abc](https://ab.co/2Um3PWI) - 'data-component'
Before this fix: `"author": null`
After this fix: `"author": "Charlotte Gore"`

[proactiveinvestors](https://bit.ly/36L8Ydp)
Before this fix: `"author": null`
After this fix: `"author": "Calum Muirhead"`

[banking](https://bit.ly/3rpyf6v)
Before this fix: `"author": "Sarah Harman Jul"`
After this fix: `"author": "Sarah Harman"`

[hcamag](https://bit.ly/36OgofI)
Before this fix: `"author": "Sarah Harman Jul"`
After this fix: `"author": "Mark Rosanes"`

[spacedaily and + 9 sites](https://bit.ly/3eGRnrf)
Before this fix: `"author": null`
After this fix: `"author": "Lucie Aubourg"`

[first](https://bit.ly/3iwTOxN) 
Before this fix:`"author": "Nick Griffin",`
After this fix: `"author": "Stan Shamu",`

[racing](https://bit.ly/3BsnRza) 
Before this fix:`"author": "Ben Sporle - @bensporle; Ben Sporle",`
After this fix: `"author": "Ben Sporle",`

[ajn](https://bit.ly/3Bs54Ea) 
Before this fix:`"author": "RABBI GARY ROBUCK July",`
After this fix: `"author": "RABBI GARY ROBUCK",`

[ESPN](https://bit.ly/3eFNJOq)  it is not totally fix, but it is better
Before this fix: `"author": "Andrew Mcglashandeputy Editor, Espncricinfo",`
After this fix: `"author": "Andrew McGlashan Deputy editor; ESPNcricinfo",`

[Probono](https://probonoaustralia.com.au/news/2021/07/charities-urged-to-create-safer-environment-for-people-of-colour/) it is not totally fix, but it is better
Before this fix: `"author": null,`
After this fix: `"author": "Luke Michael; Journalist; @Luke_Michael",`

### CODE REVIEW

**Comment by adbar** [comment]
Hi @felipehertzer, thanks for the PR. Could you please add tests for the emoji and special parts?

Please note that parsing JSON with regexes is doable but not so handy, we could also think about JSON input manipulation using the standard `json` module:

```
import json
json.loads(jsonstring)
etc.
```

**Comment by felipehertzer** [comment]
Initially I thought about using json loads but I wasn't sure about sites with invalid json, but I will switch over. Thanks

**Comment by adbar** [comment]
We could use the JSON module first and the regexes as a fallback.

This could become another PR if you're interested.

**Comment by felipehertzer** [comment]
I will left the regex for this PR, I will implement the json library in the next, is it ok?

**Comment by adbar** [comment]
Yes, let's focus on these cases first and implement a proper JSON parsing step next.

File: trafilatura/metaxpaths.py [hunk 1]
```diff
@@ -11,7 +11,7 @@
     '//*[(self::a or self::address or self::link or self::p or self::span)][@rel="author" or @id="author" or @class="author" or @itemprop="author name" or rel="me"]|//author',
     '//*[(self::a or self::div or self::p or self::span)][contains(@class, "authors") or contains(@class, "authorName") or contains(@class, "author name") or contains(@class, "author") or contains(@class, "posted-by") or contains(@itemprop, "author")]',
     '//*[(self::a or self::div or self::p or self::span)][contains(@class, "byline") or contains(@id, "author") or contains(@class, "submitted-by")]',
-    '//*[contains(@class, "author") or contains(@class, "Author") or contains(@class, "screenname")]',
+    '//*[contains(@class, "author") or contains(@class, "Author") or contains(@class, "screenname") or contains(@data-component, "Byline") or contains(@itemprop, "author")]',
```
Comment (adbar) [inline trafilatura/metaxpaths.py:14]: `contains(@itemprop, "author")` is already present two lines above

File: trafilatura/metadata.py [hunk 2]
```diff
@@ -39,7 +39,7 @@
 AUTHOR_PREFIX = re.compile(r'^([a-zäöüß]+(ed|t))? ?(by|von) ', flags=re.IGNORECASE)
 AUTHOR_REMOVE_NUMBERS = re.compile(r'\d.+?$')
 AUTHOR_REMOVE_SPECIAL = re.compile(r'[:()?*$#!%/<>{}~]')
-AUTHOR_REMOVE_PREPOSITION = re.compile(r'[^\w]+$|\b( am| on| for| at| in| to)\b\s+(.*)', flags=re.IGNORECASE)
+AUTHOR_REMOVE_PREPOSITION = re.compile(r'[^\w]+$|\b( am| on| for| at| in| to| from)\b\s+(.*)', flags=re.IGNORECASE)
```
Comment (adbar) [inline trafilatura/metadata.py:42]: Fine but the regex could be more readable in the form `...\b\s+(am|at|for|from|in|on|to)\b\s+...`

File: trafilatura/metaxpaths.py (hunk 1 above)
Comment (felipehertzer) [reply trafilatura/metaxpaths.py:14]: I put repeated because I found more elements like h3 with the itemprop, should I move it to bottom?

Comment (adbar) [reply trafilatura/metaxpaths.py:14]: I get it, that's fine, just put a comment in the code

**Comment by felipehertzer** [comment]
Hi @adbar Thanks for the feedback I changed as requested, and ran some tests and organized the xpath from more specific where we have the tags directly with the author's name to more generic tags where the code might include other elements.

**Comment by adbar** [comment]
Hi @felipehertzer, that looks good, thanks! Are you still working on it or can I merge the PR?

**Comment by felipehertzer** [comment]
Hi @adbar, yes you can merge, I have a few more sites to fix, but I need to see how we can do it. for example when the author is in the content, or when the meta tag is the name of the site but the scrapper should use the extract_author, but we can leave that for the next PR.

**Comment by adbar** [comment]
You could list the new cases in #92 or in a different thread. See also #99.
I'm merging this PR, thanks!

**Pull request merged**